### PR TITLE
Redo player_api

### PR DESCRIPTION
**Pull request by appgurueu**
- Improves code quality
- Implements #2743 as well
- Fixes `set_local_animations` call
- Removes need for workaround for changing animation speed

### CODE REVIEW

**Comment by appgurueu** [comment]
Possible compatibility breakage:

1. Mods (like `beds`) using `set_eye_offset` to counter `eye_height` of old `player_api`
2. Collisionbox changing making entering physical objects/nodes easier

File: mods/player_api/api.lua
Comment (sfan5) [inline mods/player_api/api.lua:69]: I don't know if our code style requires this, but it's done like this everywhere else.
```suggestion
	player:set_properties({textures = new_textures})
```

File: mods/player_api/init.lua
Comment (sfan5) [inline mods/player_api/init.lua:29]: was this moved? or removed?
why?

Comment (appgurueu) [reply mods/player_api/init.lua:29]: Moved to `player_api.set_model`: With each model comes it's own set of animations.

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:69]: I prefer the shorter notation. Can change though.

Comment (SmallJoker) [reply mods/player_api/api.lua:69]: Whereas this is a technically correct notation, it is not used in Minetest's code. New modders might take this code as a reference, and will be confused by this awkward special case allowed by Lua.

**Comment by paramat** [comment]
> Implements #2743 as well

There might be some good changes here, i will review later, but please do not undermine and hijack my #2743 work like this, please comment on how i can improve my PR instead. I am already working on this feature and started before you did, you saw my work and are now trying to make it redundant before i have had time to finish mine, this is offensive and irritating.

Please wait until my PR is processed one way or another then rebase your work on mine.
For this reason, i am :-1: for this PR until mine is merged or closed unmerged.

I have recently been studying the player_api mod closely over the last few weeks, because i am rewriting and improving it for use in new games. So i ask everyone, please wait until i have had time to review this PR before any possible merge.

File: mods/player_api/init.lua
Comment (sfan5) [reply mods/player_api/init.lua:29]: Ah, didn't see.

**Comment by appgurueu** [comment]
> Please do not undermine and hijack my #2743 work. I am already working on this feature and started before you did, you are now trying to make it redundant before i have had time to finish mine, this is irritating.

With all due respect for your work, the implementation of this feature takes approximately 5 minutes. I am also not satisfied with your implementation, which is why I have redone it. Differences (advantages, IMO):

No `[animation] = collisionbox/eye_height` lookup; instead per-animation collisionboxes are optional and no such lookup, instead collisionbox & eye_height are stored as subtables; higher code quality (no hacky workaround for keeping compatibility, which your lookup breaks), less code duplication; breaks compatibility if `collisionbox = {x_min, y_min, z_min, x_max, y_max, z_max}` is used as model definition.

**Comment by SmallJoker** [comment]
I don't mind if the `collisionbox` change is integrated here as long it'll be stated in the commit(s).
Overall I like this concept due to its code simplification.

**Comment by paramat** [comment]
Sorry i was irritated. You have the right to offer an alternative implementation of course, it can just be a little irritating when that happens and i have the right to get irritated by that =)
I removed my -1 above as it is not justified.

As a personal project i have been investigating how player_api could be improved for use in new games, so i really appreciate seeing work from others on this.
player_api has had the same basic form for 7 years, and it was reviewed and merged rather quickly and casually, so i have been suspecting it is not particularly good code.

Looking at your PR, it is certainly better code than my minimum-change alteration of player_api in #2743 , and fixes some other issues too, so i am closing my PR.
The Player API is the part of MTG most likely to be used in new games, so i think it is a good idea to get this mod into a good state despite MTG being feature-frozen.

File: game_api.txt
Comment (paramat) [inline game_api.txt:448]: It is convention to include the argument brackets when referring to a function.

File: mods/player_api/api.lua
Comment (paramat) [inline mods/player_api/api.lua:61]: `player_data` is local to the function, surely this needs to write data to `players[player_name]`?

File: mods/player_api/api.lua
Comment (paramat) [inline mods/player_api/api.lua:68]: `player_data` is local to the function, surely this needs to write data to `players[player_name]`?

File: mods/player_api/api.lua
Comment (paramat) [inline mods/player_api/api.lua:83]: `player_data` is local to the function, surely this needs to write data to `players[player_name]`?

File: mods/player_api/api.lua
Comment (paramat) [inline mods/player_api/api.lua:86]: Ideally, this should only be executed if the collisionbox or eye_height changes (as in my PR). With the MTG player model, these only change when sitting or laying.

File: mods/player_api/api.lua
Comment (paramat) [inline mods/player_api/api.lua:133]: Why is this no longer needed?
I admit i actually do not understand why this code block is needed.

Comment (paramat) [reply mods/player_api/api.lua:133]: Ah, perhaps this is what you mean by 'Removes need for workaround for changing animation speed' ?
In https://github.com/minetest/minetest_game/commit/4c0c7ae146bb57f56eb03e804853db73d31ed957#diff-ee5a751279baa5e37734fc0c1e75daf3 the old comment for this was:
`-- Refresh player animation below if sneak state changed`

File: game_api.txt
Comment (appgurueu) [reply game_api.txt:448]: `f()` refers to a call to the function without any arguments. I might decide to use `f(...)`, but I sure won't use `f()`.

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:61]: Not at all. `players[player_name]` is a table. `player_data.model` is the same as `players[player_name].model`.

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:68]: See above.

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:83]: See above.

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:86]: Engine responsibility. See minetest/minetest#9934

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:133]: > Ah, perhaps this is what you mean by 'Removes need for workaround for changing animation speed' ?

Yes.

File: game_api.txt
Comment (paramat) [reply game_api.txt:448]: Ok, this is fine as it is, f(...) seems worse to me.

File: mods/player_api/api.lua
Comment (paramat) [reply mods/player_api/api.lua:61]: Yes, i just remembered, tables in Lua contain references to values, so i assume altering a copy of a table alters the original table.

File: mods/player_api/api.lua
Comment (paramat) [reply mods/player_api/api.lua:86]: Hmm, ideally yes, but the engine does not do that yet, and with our slow dev may not do so for years.
I would like to avoid setting properties and sending packets every time a player starts or stops walking or mining, which is very often.
Please could you attempt this?
But i will not insist, someone else could add that later.

Comment (appgurueu) [reply mods/player_api/api.lua:86]: I see three options:

1. Don't do it. Higher code quality, unnecessary packets in practice. Issue can be fixed better by the engine.
2. Do it, to avoid those packets, annotate as hack.
    1. Compare object properties, requires get_properties call and relatively cheap comparison, will override values changed by other mods
    2. Compare cached values, faster, neater, won't override values if changed by other mods (probably desirable)

We might want to settle on 2.2.

Comment (paramat) [reply mods/player_api/api.lua:86]: I think it is best we wait for input from another core dev concerning what we do here.

Comment (sfan5) [reply mods/player_api/api.lua:86]: I suggest 2., whether i. or ii. doesn't really matter. Longterm the engine needs a proper `player_monoids`-esque solution to deal with multiple mods want to set something.

Comment (lhofhansl) [reply mods/player_api/api.lua:86]: FWIW. I agree with (2). Yes, it should be engine optimizing this, but until we have that, let's have mods be smart.
(See my NSSM change to avoid hammering the server with unnecessary node updates... Even though the server should avoid applying no-op changes).

Comment (paramat) [reply mods/player_api/api.lua:86]: In case it is not obvious, i support 2 also.

**Comment by appgurueu** [comment]
So, my solution looks like this:

1. During model registration:
    1. Delete values equal to the model default values
    2. Use same references for equal collisionboxes
2. This allows us to simply check (during set_animation):
    1. `prev_anim.eye_height == anim.eye_height`
    2. `prev_anim.collisionbox == anim.collisionbox`

Performance could be increased by caching some sort of property combination ID which could then be compared in `set_animation` (single comparison instead of double). Thoughts? Should the "model registration" part be removed and rather all comparisons be done in `set_animation` (could be seen as increased code quality, at the expense of performance) ?

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua]: See https://github.com/minetest/minetest_game/pull/2745#issuecomment-706591210

**Comment by paramat** [comment]
In my PR i discovered you cannot do boolean '==' or '~=' between the collision tables, so i ended up using `table.concat()` to get values that worked:
`if table.concat(new_cbox) == table.concat(old_cbox) then`
What do you think of that approach? It avoids mod code that loops through table values to compare them, but may actually be lower performance. It also seems a little crude.

File: game_api.txt
Comment (paramat) [inline game_api.txt:487]: Why are single-value collisionbox and eye_height still required in the def when the new approach is to define them in the animations?
Seems messy and confusing, i think these should only be defined in animations.

File: game_api.txt
Comment (paramat) [inline game_api.txt:479]: I do not understand why these are documented as 'default collisionbox' and 'default eye height', surely these are not the defaults. How about:
`collisionbox = <anim collisionbox>, eye_height = <anim eye height>`

Comment (paramat) [reply game_api.txt:479]: `<anim_name>`
seems more suitable for 'insert your value here', as in previous code.

**Comment by paramat** [comment]
> During model registration:
Delete values equal to the model default values

I dislike that, i think it is less confusing and more useful to always store the value.
Besides, i also dislike how you are requiring the def contains default values as well as the values defined in the animation.
I think defaults should not exist, that adds more complexity and confusion for no reason, there should only be values defined in every animation.

Performance is not critical, as 'set animation' is only executed approx. once every second.

**Comment by paramat** [comment]
I may be wrong, but i am not sure this avoids the need to do a boolean comparison of 2 tables, which is not possible.
This may be clever code, but i am therefore struggling to understand it.

Overall, i really like parts of this PR, and really dislike other parts. I feel it might be 'too clever' and so neglecting simplicity. It seems over-concerned with performance, which is not critical.
> Thoughts? Should the "model registration" part be removed and rather all comparisons be done in set_animation (could be seen as increased code quality, at the expense of performance) ?

Yes please.

File: game_api.txt
Comment (appgurueu) [reply game_api.txt:487]: These are default values if there are no `eye_height` or `collisionbox` values provided by animations. This allows for increased code quality (no repetition of these values) and backwards compatibility.

File: game_api.txt
Comment (appgurueu) [reply game_api.txt:479]: Hmm, my thought was that if you do `key = <what_the_value_should_be>` it's not the same as `key = default value`. I wanted to make clear that there are defaults, but this is probably misleading.

**Comment by appgurueu** [comment]
> Please also remove the default collisionbox and eye_height fields from the def, if possible.

Hmm. I won't remove those fields, but I can choose between:

1. Setting values equal to the defaults to `nil` or
2. Setting all `nil` values to the defaults

In order to keep backwards compatibility.

**Comment by appgurueu** [comment]
https://github.com/minetest/minetest_game/pull/2745/commits/78d358c6d7e956c0b7eba8f8ecc19b4e9b091b1b sets all defaults on registration (slightly better for performance and probably readability). Note that I never required the model defaults, I just kept them.

It also uses an ID (`_equals`) set to the name of an animation with same properties to check for animation property equality. Highest performance, IMO more understandable than only doing the ID workaround for the collisionbox.

**Comment by appgurueu** [comment]
Newest commit ensures backwards compatibility by only logging a warning instead of crashing if the player is offline.

File: mods/player_api/api.lua
Comment (SmallJoker) [inline mods/player_api/api.lua:30]: This code takes a lot time to read. Please comment the actions done here and why *exactly* the `_equals` copy is needed here. Even after reading the discussion I don't understand how these cannot be taken directly using `table.copy`.

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:52]: It is human readable, as a warning is logged (code below). Using the error flow is required to avoid code duplication. If you don't like this, I can also throw an error table and do an ID comparison.

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:30]: I guess you're referring to the loop? It basically finds another animation with equal properties and then takes the key of that animation.

File: mods/player_api/api.lua
Comment (SmallJoker) [reply mods/player_api/api.lua:52]: ¯\_(ツ)_/¯ well then. good enough I guess...

File: mods/player_api/api.lua
Comment (appgurueu) [reply mods/player_api/api.lua:30]: The purpose of the loop is to sort animations into "property classes".

All animations in a "property class" have equal properties (equal eye height and collisionbox), so there is no need to call `set_properties` if the previous animation is in the same property class (`previous_anim_equals == anim._equals`).

Those "property classes" use the name of the "first" (in "pairs" traversal order) animation with those properties.

For example, the "property classes" of the default model might be: `{ stand = { ..., _equals = "stand" }, lay = { ..., _equals = "lay" }, walk = { ..., _equals = "stand" }, walk_mine = { ..., _equals = "stand" }, sit = { _equals = "sit" } }`

"stand", "walk" and "walk_mine" all share one _equals value, "sit" and "lay" have their own _equals values

**Comment by appgurueu** [comment]
Ready for merge.

**Comment by sfan5** [comment]
Bug: For the first time in a game session sleeping in a bed will not set the model.

**Comment by appgurueu** [comment]
> Bug: For the first time in a game session sleeping in a bed will not set the model.
> 
> This doesn't seem to be caused by this PR however...

It is not. It depends on whether you enter the bed in first-person mode, AFAIK.

**Comment by appgurueu** [comment]
Bump. Merge conflicts resolved.

**Review by sfan5 — APPROVED** [review]
Tested, works.

**Comment by appgurueu** [comment]
Last commit is untested and effectively implements https://github.com/minetest/minetest_game/pull/2751

File: mods/player_api/api.lua
Comment (Zughy) [inline mods/player_api/api.lua:113]: I'd invert `index` and `texture`, to make it coherent with `set_textures`

Comment (appgurueu) [reply mods/player_api/api.lua:113]: Imagining this as a table, I consider `playertextures[index] = texture` more intuitive.

**Review by sfan5 — APPROVED** [review]
Appears to work fine

**Comment by sfan5** [comment]
I am still very curious why changing the player model (when entering a bed) does not work in third-person mode. This detail should be mostly invisible to the game.

**Pull request merged**
